File: Graphs/dfs.py
class node:
    def __init__(self,n):
        self.neighbors = []
        self.visited = False
        self.name = n 

def dfs(start):
    stack = []
    stack.append(start)

    while (stack != []):
        top = stack.pop()

        if (top.visited == False):

            top.visited = True
            print(top.name)
            #In this part in the termination condition

        n = top.neighbors
        for i in n:
            stack.append(i)



a = node("a")
b = node("b")
c = node("c")
d = node("d")


a.neighbors = [b,c]
c.neighbors = [d]
dfs(a)
        

File: Graphs/dfs.cpp
#include <iostream>
#include <vector>
#include <stack>
#include <string>

using namespace std;

typedef struct node {
    string val;
    bool visited = false;
    vector < node >  neighbors ;
} node;


typedef vector < node >  list_nodes;



inline void dfs(node start){
    stack<node> s;
    s.push(start);
    while(s.empty() == false){
        node top = s.top();
        s.pop(); 

        if (top.visited == false){
            top.visited  = true;
            cout << "Node " << top.val << " visited" << endl;
        }

        list_nodes n = top.neighbors;
        cout << "\tNum of neighbors " << n.size() << endl;
        for (node a: n){
            s.push(a);
        }

    }
}


int main(){
 
    node a,b,c,d ;
    a.val = "a";
    b.val = "b";
    c.val = "c";
    d.val = "d";

    list_nodes la(2), lc(1);
    lc[0] = d;
    la[0] = b;
    la[1] = c;
    

    c.neighbors = lc;
    a.neighbors = la;
    

    dfs(a);
    return 0; 

}
